### PR TITLE
NIFI-10273: added support for files larger than 8.5GB to MergeContent…

### PR DESCRIPTION
**Pull request by michael81877**
… (tar) processor

# Summary

[NIFI-10273](https://issues.apache.org/jira/browse/NIFI-10273)

Adds support for the MergeContent processor to handle individual files larger than 8.5GB when Tar archiving them. I attempted to unit test this, but due to how the NiFi mock framework works, every flowfile is read into memory in full even if that FlowFile is backed by a `NullInputStream`. I couldn't find a nice alternative to verify with a unit test.

So to verify i deployed this processor and tested with the following setup, with the `MergeContent` processor configured to tar archive the bins:

`GenerateFlowFile` --- 9GB file ---> `MergeContent` --> `LogAttribute`

Prior to this fix, the MergeContent processor would complain with the following error:
```
java.lang.RuntimeException: entry size '<FlowFileSize>' is too big (> 8589934591)
```

A similar fix was made in a different project: https://github.com/DataConservancy/dcs-packaging-tool/pull/36/files

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 8
  - [x] JDK 11
  - [x] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by exceptionfactory** [comment]
Thanks for the background and runtime testing on this issue @michael81877.

The change seem straightforward, and I agree that attempting to build a unit test for file that large is not helpful.

Reviewing the [Commons Compress TarArchiveOutputStream](https://commons.apache.org/proper/commons-compress/apidocs/org/apache/commons/compress/archivers/tar/TarArchiveOutputStream.html#BIGNUMBER_POSIX) documentation and source code, setting the `BIGNUMBER_POSIX` mode unconditionally writes the TAR entry headers using a [different approach](https://pubs.opengroup.org/onlinepubs/009695399/utilities/pax.html#tag_04_100_13_03).

According to references, it seems like this format has been around long enough that it is supported in modern TAR utilities. With that being said, there is the outside possibility that this change could cause complications with some legacy system.

The `TarMerge.merge()` method has a reference to all the FlowFiles that will be written, so it seems possible to calculate the total size prior to setting the mode. By calculating the total and checking it against a value, it would be possible to set this flag only when the output is expected to exceed the maximum size. What do you think of that approach?

**Comment by michael81877** [comment]
> the TAR entry headers usin

Thanks for the input @exceptionfactory ! Do you think that having the processor emit tarred FlowFiles with a potential mix of entry headers depending on input FlowFile sizes is a less desirable behavior? What about adding a new property to set "large file mode" so the operator opts in to the behavior for all FlowFiles or none of them? I'm good with either approach - just trying to determine the least intrusive approach.

Thanks again for the review!

**Comment by michael81877** [comment]
> Thanks for the background and runtime testing on this issue @michael81877.
> 
> The change seem straightforward, and I agree that attempting to build a unit test for file that large is not helpful.
> 
> Reviewing the [Commons Compress TarArchiveOutputStream](https://commons.apache.org/proper/commons-compress/apidocs/org/apache/commons/compress/archivers/tar/TarArchiveOutputStream.html#BIGNUMBER_POSIX) documentation and source code, setting the `BIGNUMBER_POSIX` mode unconditionally writes the TAR entry headers using a [different approach](https://pubs.opengroup.org/onlinepubs/009695399/utilities/pax.html#tag_04_100_13_03).
> 
> According to references, it seems like this format has been around long enough that it is supported in modern TAR utilities. With that being said, there is the outside possibility that this change could cause complications with some legacy system.
> 
> The `TarMerge.merge()` method has a reference to all the FlowFiles that will be written, so it seems possible to calculate the total size prior to setting the mode. By calculating the total and checking it against a value, it would be possible to set this flag only when the output is expected to exceed the maximum size. What do you think of that approach?

**Comment by michael81877** [comment]
That sounds reasonable to me. I will add some logic to conditionally set the big number mode only if the FlowFile is larger than the [max tar entry size](https://commons.apache.org/proper/commons-compress/apidocs/src-html/org/apache/commons/compress/archivers/tar/TarConstants.html#line.94).

Thanks!

**Comment by exceptionfactory** [comment]
@michael81877 Just checking in to see if you are making progress on the changes, glad to help if you run into any issues.

**Comment by michael81877** [comment]
> @michael81877 Just checking in to see if you are making progress on the changes, glad to help if you run into any issues.

Sorry, my apologies! I've been on leave for a bit. I'll make this simple change this week! Thank you for following up 🙂

**Comment by michael81877** [comment]
@exceptionfactory : So i think there's an issue with the conditional mode approach which i'd overlooked when you suggested it. The core of this issue is that `MergeContent` cannot handle when _any single tar entry_ is larger than 8.5GB. It has no issue creating tar archives which end up being larger than 8.5GB in total. The output mode needs to be set on the entire tar archive, so i feel this is an all or nothing thing.

Perhaps we could compute whether any one tar entry (e.g. FlowFile) is larger than 8.5GB and if that's the case then we set the `bigNumberMode()` and otherwise leave the implementation as is. I believe the only other alternative is to add a processor property to have it always set the `bigNumberMode()`.

Thoughts?

**Comment by exceptionfactory** [comment]
Thanks for the update @michael81877. The `merge()` method calls `Bin.getContents()`, which returns all the FlowFiles. It should be possible to check the size of the FlowFiles and if any one is over the limit, set a flag to enable big number mode.

**Comment by michael81877** [comment]
> Thanks for the update @michael81877. The `merge()` method calls `Bin.getContents()`, which returns all the FlowFiles. It should be possible to check the size of the FlowFiles and if any one is over the limit, set a flag to enable big number mode.

Pushed a commit to do this. Let me know if you see any issues!

File: nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeContent.java [hunk 1]
```diff
@@ -827,6 +833,17 @@ public void process(final OutputStream rawOut) throws IOException {
...
 
+        private long getMaxEntrySize(List<FlowFile> contents) {
+            Optional<FlowFile> ffOpt = contents.stream()
+                .parallel()
+                .max(Comparator.comparingLong(ff -> ff.getSize()));
+            if (ffOpt.isPresent()) {
+                return ffOpt.get().getSize();
+            } else {
+                return 0L;
+            }
```
[the remark this answers is not in the document]
Comment (michael81877) [reply nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/MergeContent.java:844]: Ah that's a good shout. Changing that!

Will `mapToLong()`, `max()`, and return `longVal.orElse(0)`

**Pull request closed without merging**